### DISCUSSION TITLE
is there any reason we don't have separate skinnable early/late 50/100s?

### DISCUSSION THREAD

**Discussion by tovc** [ppy/osu#31248] [Ideas] [answered]
It seems like a relatively obvious request, but I couldn't find a discussion regarding it, but I'm sorry if there is one

We already have a hit error bar that achieves some of this functionality, but it requires you to look away from your gameplay, which often causes misses. The only time I find myself using the hit error bar successfully, is as I'm going to restart a map. In that vein, I think it'd be much easier to correct your timing as you're playing if it was communicated through the judgements. Not only would this allow you to keep your eyes a lot closer to the relevant gameplay, it'd also allow you to know exactly which circles were hit early/late, rather than having them buried in a forest of other judgements, on the hit error bar

To be clear, I *do* think that the hit error bar is a useful thing to have, but it's just not well suited to telling you whether or not you're hitting early or late during gameplay. I know one response is gonna be something along the lines of 'just figure it out by feel', and while that definitely works at lower ODs, it becomes rather difficult at higher ODs. 'Feeling it out' has been the way to do it for the last 17 years, and I think we can do better

**Comment by peppy** [comment, marked as answer]
No reason. We'll probably add something like this in the future.

**Reply by peppy** [reply]
Tracking https://github.com/ppy/osu/issues/36845

**Comment by GXboy12345** [comment]
Definitely something I'd like to see too...

It could be as simple as allowing an optional skin expansion, with each judgment's graphic having an early/late variant displayed depending on the delta of the last hit object.

An example of a potential graphic, based on the "triangles" lazer skin:
<img width="285" alt="408970891-b3331eb4-ec20-46dc-a29b-851c94930d37" src="https://github.com/user-attachments/assets/90bf6e00-5414-49ec-a183-74aa145f2c55" />

**Comment by Stalzak** [comment]
Adding some UX/UI design ideas to this discussion!

The latest Taiko no Tatsujin game experiments with displaying icons next to the judgement text in training mode. Icons are useful, since they take way less space than writing out a long and language-dependent "Early" or "Late". Depending on the final design, these icons could be integrated without adding too much visual noise- assuming that both icons are recognizable enough at a glance, or placed on different sides at least. (Reference and rough mockup with lazer's hit error meter's icons.)

<img width="492" height="164" alt="image" src="https://github.com/user-attachments/assets/1b55d02f-df46-47b3-b5c2-ddc1c70671e9" />

Icons like that could act as two skinnable elements anchored to the left or right side of judgement assets, making them optional and ready to use across all rulesets.

Another and maybe additional way to convey this information could also be through animations: in ADOFAI, the different early and late judgements are animated to pivot respectively to the left or the right. Since movement perception is much faster in human vision than recognizing and reading a static graphic, this might make identifying early and late inputs much more intuitive at high levels. (I'm not skilled enough yet to craft an mockup in-engine, but here's a capture of ADOFAI's animations for reference.)

![ezgif-383353e93c800243](https://github.com/user-attachments/assets/2361404e-e0e1-44d6-ade4-2ebc6b0f70d8)

Edit: animating the whole judgement graphic might be overkill and too visually distracting, but applying a quick movement and/or rotation of early/late icons could be an option.

**Reply by peppy** [reply]
I really like the turtle/rabbit iconography.

**Reply by GXboy12345** [reply]
Specifically, TnT:RF implements these icons as part of their training mode of free play, which also (notably) allows the user to see their early/late totals by segment on the endscreen.

If this feature ever gets implemented (fingers crossed), I want to throw my vote into the hat and ask that late be marked with red and early be marked with blue!

**Comment by koh-hinooka** [comment]
As a player who mainly plays osu!mania, I'd like to add four requirement ideas for the early/late indicator tracked in #36845.

**Context**
On high-OD mania charts, my eyes stay fixed around the centre of the playfield, where the judgement text and combo counter are. The hit error bar can't be read without looking away, and mapping "left is early / right is late" by position is hard to do at a glance. An indicator shown right next to the judgement would let me correct my timing without breaking my play.

**Suggestions**

1. **An option to show the offset in milliseconds** (e.g. `EARLY -12ms` / `LATE +8ms`)
   Knowing the amount as well as the direction makes it possible to tell "slightly early" apart from "very early". As mentioned in this thread, PR #7310 was closed because three states weren't enough, so a numeric display could be one way to address that.

2. **An option to hide the indicator for the highest judgement (PERFECT in mania)**
   Showing it on every judgement adds a lot of on-screen noise. Showing it only on judgements that actually need correcting keeps the playfield readable.

3. **A skinnable element independent of the judgement sprite**
   If it were a HUD (or per-ruleset) component that can be positioned and scaled in the skin editor, skins could place it wherever suits their layout, such as directly below the judgement. For legacy skins, being able to render the digits with the existing score font sprites (`ScorePrefix` etc.) would keep it consistent with the rest of the skin.

4. **Early and late distinguishable by text or icon**
   A `+` / `-` sign alone is hard to read at a glance. Being able to choose a representation whose direction is obvious, such as `EARLY` / `LATE` text or the rabbit/turtle icons suggested in this thread, would help a lot.

I'm happy to help with testing and feedback on mania once work on this starts.
